### PR TITLE
Fix a bug that OpenTelemetry span is not closed in right order

### PR DESCRIPTION
**Pull request by FrankChen021**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fixes a bug found by stress test. The original report is at: https://github.com/ClickHouse/ClickHouse/pull/39010#issuecomment-1243592385

And the detailed cause analysis is at here: https://github.com/ClickHouse/ClickHouse/pull/39010#issuecomment-1243870099 and https://github.com/ClickHouse/ClickHouse/pull/39010#issuecomment-1244972762

It's very hard to produce such problem, it needs to meet following 3 conditions
1. OpenTelemetry tracing is enabled for queries
2. must send 'SHOW PROCESSLIST' query via clickhouse-client
3. before getting query result from the server, the client disconnect the TCP connection on which query is sent

The solution is to make sure the query_span created in `executeQuery` is closed in the `onFinish` callback. 
So I inverted the if condition to include all things that needs to be done, and put the code related to query span to the last.

If there's exception thrown before the `span->finish` is called, caller of `executeQuery` makes sure that `onException` callback will be called(current behavior). `onException` will call the `span->finish` which also ensure the span is closed.

Also I optimized the `SpanHolder::finish()` to make sure it will only be called only once.

cc @qoega @rschu1ze

### CODE REVIEW

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    AND   attribute['clickhouse.query_id'] = '${1}'
+    Format JSONEachRow
+    ;"
+}
+
+#
+# Set up
+#
+${CLICKHOUSE_CLIENT} --database=${CLICKHOUSE_DATABASE} -nq "
+DROP   TABLE IF EXISTS opentelemetry_test;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:53]: Cosmetic: `DROP TABLE` (no space in between)

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# This function takes 2 arguments:
+# $1 - query
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:8]: `$1` looks to me like a query_id and `$2` like a query

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+# $2 - trace id
+function executeQuery()
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:10]: `execute_query()`

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+{
+  ${CLICKHOUSE_CLIENT} --database=${CLICKHOUSE_DATABASE} --opentelemetry_start_trace_probability=1 --query_id $1 -nq "
+      ${2}
+  "
+}
+
+# For some quries, it's not able to know how many bytes/rows are read when tests are executed on CI,
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:17]: quries --> queries, able --> possible

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# This function takes 2 arguments:
+# $1 - query
+# $2 - trace id
+function executeQuery()
+{
+  ${CLICKHOUSE_CLIENT} --database=${CLICKHOUSE_DATABASE} --opentelemetry_start_trace_probability=1 --query_id $1 -nq "
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:12]: Btw., nice trick with passing a query_id explicitly. Didn't know this is possible.

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+function check_query_span()
+{
+${CLICKHOUSE_CLIENT} -nq "
+    SYSTEM FLUSH LOGS;
+    SELECT attribute['db.statement']             as query,
+           attribute['clickhouse.read_rows']     as read_rows,
+           attribute['clickhouse.read_bytes']    as read_bytes,
+           attribute['clickhouse.written_rows']  as written_rows,
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:39]: The number of read and written bytes seems like an internal detail which is rather fragile and may break with future (unrelated!) commits. Suggest to not check it.

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh (hunk 5 above)
Comment (FrankChen021) [reply tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:12]: It works fine. The test case will find related span logs based on the given query id.

File: tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh (hunk 6 above)
Comment (FrankChen021) [reply tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:39]: OK, Let's keep rows only.

**Comment by FrankChen021** [comment]
Stateless tests (debug, s3 storage) [2/3] — fail: 3

These 3 tests fails, and have not found any log related to opentelemetry

- 01085_max_distributed_connections | FAIL | 600.02
- 01516_drop_table_stress_long
- 01195_formats_diagnostic_info

Stateless tests (release) — fail: 1, passed: 4447, skipped: 1

I don't find the fail case in the details

**Comment by FrankChen021** [comment]
Thank you @rschu1ze

**Pull request merged**